WordPress/Classes/ViewRelated/NUX/SigninEmailViewController.swift

```diff
@@ -286,49 +285,89 @@ import WordPressShared
-
-            } else if restrictSigninToWPCom {
-                // When restricted, show a prompt then let the user enter a new username.
-                SigninHelpers.promptForWPComReservedUsername(emailOrUsername, callback: {
-                    self.loginFields.username = ""
-                    self.emailTextField.text = ""
-                    self.emailTextField.becomeFirstResponder()
-                })
-
-            } else {
-                // Switch to the signin flow when not restricted.
-                signinToSelfHostedSite()
+        if emailOrUsername.isValidEmail() {
+            validate(email: emailOrUsername)
+        } else {
+            validate(username: emailOrUsername)
+        }
+    }
+
+    private func validate(email: String) {
+        let service = AccountService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        configureViewLoading(true)
+
+        service.isEmailAvailable(email,
+                                 success: { [weak self] (available: Bool) in
+                                    self?.configureViewLoading(false)
+                                    if (available) {
+                                        // No matching email address found so treat this as a
+                                        // self-hosted sign in.
+                                        self?.signinToSelfHostedSite()
+                                    } else {
+                                        self?.requestLink()
+                                    }
+            },
+                                 failure: { [weak self] (error: Error) in
+                                    DDLogSwift.logError(error.localizedDescription)
+                                    guard let strongSelf = self else {
+                                        return
+                                    }
+                                    strongSelf.configureViewLoading(false)
+                                    strongSelf.displayError(error as NSError, sourceTag: strongSelf.sourceTag)
+        })
+    }
+
+    private func validate(username: String) {
+        if SigninHelpers.isWPComDomain(username) {
+            signinWithWPComDomain(username)
+        } else if !SigninHelpers.isUsernameReserved(username) {
+            signinWithUsernamePassword()
+        } else if restrictSigninToWPCom {
+            // When restricted, show a prompt then let the user enter a new username.
+            SigninHelpers.promptForWPComReservedUsername(username, callback: {
+                self.loginFields.username = ""
+                self.emailTextField.text = ""
+                self.emailTextField.becomeFirstResponder()
+            })
+        } else {
+            // Switch to the signin flow when not restricted.
+            signinToSelfHostedSite()
+        }
+    }
+
+    private func signinWithWPComDomain(_ domain: String) {
+        func showFailureError() {
+            weak var weakSelf = self
```

Comment:
I'm not sure, but I think since `self` is referenced inside the nested function it will always be a strong reference. 
I think we'd want to move the `weakSelf` declaration outside the nested function... but yah... not sure and what I'm reading in the docs isn't clear. 

Alternatively, the question becomes moot by tweaking the syntax from a nested function to an assigned closure like so:
```
let showFailureError = { [weak self] in
            guard let strongSelf = self else {
                return
            }
            strongSelf.displayErrorMessage(NSLocalizedString("Please enter a valid username or email address", comment: "A short prompt asking the user to properly fill out all login fields."))
        }
```

Thoughts?